### PR TITLE
Issue 2398 make tests inline images

### PR DESCRIPTION
**Pull request by michael-volynets**
Issue: #2398
Time Spent: about 10hrs

It takes much more than I expected,
At the moment, I only finished writing test for checking if images aren't present in a footer, and if images render correctly when saving/rendering draft.
The best way to test drafts I found is to try to create a draft with an image and render that draft in one test so we can make sure that everything is working. But while working on it I faced a lot of issues.
The first one was that it didn't work (I mean I couldn't have seen any rich text elements (e.g. link, images)) so I fixed that in my first commit.
The next issue is I have to somehow store these newly created drafts in our `mock-data.ts` file, it gave me some troubles to map the data from the request to the `GmailMsg` / `DraftMsg`.
The third issue I had to deal with is there is no way to paste `image/*` using a clipboard in puppeteer, so I added `<input type="hidden" />` only in `debug` when I put a base64 test of image and squire adds that image to the textbox.

When I finish writing tests for encrypted/signed messages with images I will make this PR ready for review

### CODE REVIEW

File: extension/chrome/elements/composer/composer-send-btn-popover.ts [hunk 1]
```diff
@@ -27,7 +27,7 @@ export class ComposerSendBtnPopover extends ComposerComponent {
     this.choices.richText = await this.richTextUserChoiceRetrieve();
     for (const key of Object.keys(popoverItems)) {
       const popoverOpt = key as PopoverOpt;
-      if (popoverOpt === 'richText' && Catch.browser().name !== 'firefox') {
+      if (popoverOpt === 'richText' && (!this.view.debug && Catch.browser().name !== 'firefox')) {
```
Comment (tomholub) [inline extension/chrome/elements/composer/composer-send-btn-popover.ts]: The brackets are not useful since you are not combining `&&` with `||`

File: extension/chrome/elements/composer/composer-input.ts [hunk 2]
```diff
@@ -187,6 +190,15 @@ export class ComposerInput extends ComposerComponent {
     return result;
   }
 
+  // We need this method to test imagees in drafts because we can't paste them dirctly in tests.
```
Comment (tomholub) [inline extension/chrome/elements/composer/composer-input.ts]: A typo above.

It's workable as a workaround but I'd hope we could find some other way to achieve this. If not with puppeteer, then using OS level utilities.

File: test/source/tests/tests/compose.ts [hunk 3]
```diff
@@ -525,6 +526,23 @@ export const defineComposeTests = (testVariant: TestVariant, testWithNewBrowser:
...
+        $('[data-test=action-insert-image]').val('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABkCAYAAABw4pVUAAAAnElEQVR42u3RAQ0AAAgDIE1u9FvDOahAVzLFGS1ECEKEIEQIQoQgRIgQIQgRghAhCBGCECEIQYgQhAhBiBCECEEIQoQgRAhChCBECEIQIgQhQhAiBCFCEIIQIQgRghAhCBGCEIQIQYgQhAhBiBCEIEQIQoQgRAhChCAEIUIQIgQhQhAiBCEIEYIQIQgRghAhCBEiRAhChCBECEK+W3uw+TnWoJc/AAAAAElFTkSuQmCC')
+          .click();
+      });
+      await ComposePageRecipe.waitWhenDraftIsSaved(composePage);
+      await composePage.close();
+      const appendUrl = 'draftId=draft_testsavingdraftsinmock';
+      composePage = await ComposePageRecipe.openStandalone(t, browser, 'compatibility', { appendUrl });
+      const body = await composePage.waitAny('@input-body');
+      const image = await body.$eval('img', el => el.getAttribute('src'));
+      expect(image).to.not.be.empty;
```
Comment (tomholub) [inline test/source/tests/tests/compose.ts]: You could actually compare the image src to equal

Comment (michael-volynets) [reply test/source/tests/tests/compose.ts]: Yes, agree, it's better.
Thanks!

**Comment by michael-volynets** [comment]
+11h 30m,

Something wrong with the tests, they pass locally

**Comment by tomholub** [comment]
Mock reported this error. This looks like a bug in the mock code.
```
500 POST /upload/gmail/v1/users/me/messages/send?uploadType=multipart | {"error":{"message":"Cannot read property 'length' of undefined"}}
```

To fix this, make mock return:
```
{"error":{"message":"Cannot read property 'length' of undefined", stack: "..."}}
```
Then you can diagnose.

And also have a look at all `.length` in mock code.

I should not have to tell you this. It cannot be my sole responsibility to make tests work well in the CI. If a test works locally but not in the CI, investigate.

File: extension/chrome/elements/composer/composer-draft.ts [hunk 4]
```diff
@@ -64,12 +65,13 @@ export class ComposerDraft extends ComposerComponent {
         return;
       }
       const parsedMsg = await Mime.decode(Buf.fromBase64UrlStr(draftGetRes.message.raw!));
-      const armored = PgpArmor.clip(parsedMsg.text || Xss.htmlSanitizeAndStripAllTags(parsedMsg.html || '', '\n') || '');
+      const encryptedMsg = parsedMsg.atts.find(att => att.treatAs() === 'encryptedMsg')?.getData().toUtfStr();
```
Comment (tomholub) [inline extension/chrome/elements/composer/composer-draft.ts]: Here we should re-use the same mechanism that we use when decrypting messages in `pgp_blocks`, for consistency.

File: extension/chrome/elements/composer/composer-draft.ts [hunk 5]
```diff
@@ -64,12 +65,13 @@ export class ComposerDraft extends ComposerComponent {
...
+      const armored = PgpArmor.clip(parsedMsg.text || Xss.htmlSanitizeAndStripAllTags(parsedMsg.html || '', '\n') || '') || encryptedMsg;
       if (!armored) {
         await this.abortAndRenderReplyMsgComposeTableIfIsReplyBox('!armored');
         return;
       }
-      await this.decryptAndRenderDraft(armored, parsedMsg);
+      await this.decryptAndRenderDraft(armored, parsedMsg, armored === encryptedMsg);
```
Comment (tomholub) [inline extension/chrome/elements/composer/composer-draft.ts]: Therefore this should not be necessary

File: extension/chrome/elements/composer/composer-draft.ts [hunk 6]
```diff
@@ -95,35 +97,38 @@ export class ComposerDraft extends ComposerComponent {
       this.currentlySavingDraft = true;
```
Comment (tomholub) [inline extension/chrome/elements/composer/composer-draft.ts]: This method has gotten quite large, will see if it can be simplified

File: extension/chrome/elements/composer/composer-draft.ts [hunk 7]
```diff
@@ -188,7 +194,23 @@ export class ComposerDraft extends ComposerComponent {
           this.composer.S.cached('input_subject').val(headers.subject);
         }
         this.composer.S.cached('prompt').css({ display: 'none' });
-        this.composer.input.inputTextHtmlSetSafely(Xss.escape(result.content.toUtfStr()).replace(/\n/g, '<br>'));
+        let decryptedContent: string;
+        if (!Mime.resemblesMsg(result.content)) {
```
Comment (tomholub) [inline extension/chrome/elements/composer/composer-draft.ts]: This looks very similar to code from other places, should abstract it

File: extension/chrome/elements/composer/composer-draft.ts [hunk 8]
```diff
@@ -197,7 +219,7 @@ export class ComposerDraft extends ComposerComponent {
       }
     } else {
       await this.renderPPDialogAndWaitWhenPPEntered();
-      await this.decryptAndRenderDraft(encryptedArmoredDraft, headers);
+      await this.decryptAndRenderDraft(encryptedArmoredDraft, headers, isRichText);
```
Comment (tomholub) [inline extension/chrome/elements/composer/composer-draft.ts]: Hopefully this won't be needed

File: extension/chrome/elements/composer/composer-footer.ts [hunk 9]
```diff
@@ -41,9 +41,9 @@ export class ComposerFooter extends ComposerComponent {
       return '';
     }
     if (/^[*-_=+#~ ]+$/.test(footerFirstLine)) {
-      return `<br>${sanitizedHtmlFooter}`;  // first line of footer is already a footer separator, made of special characters
+      return sanitizedHtmlFooter;  // first line of footer is already a footer separator, made of special characters
     }
-    return `<br>--<br>${sanitizedHtmlFooter}`; // create a custom footer separator
+    return `--<br>${sanitizedHtmlFooter}`; // create a custom footer separator
```
Comment (tomholub) [inline extension/chrome/elements/composer/composer-footer.ts]: Looks ok if the `<br>` was moved elsewhere

Also have to see if I have intended to always use `<br / >` instead, I vaguely remember wanting to do that but forgot.

File: extension/chrome/elements/composer/composer-send-btn-popover.ts [hunk 10]
```diff
@@ -12,23 +12,23 @@ import { Xss } from '../../../js/common/platform/xss.js';
...
       sign: { text: 'Sign message', iconPath: '/img/svgs/signature-gray.svg' },
     };
-    this.choices.richText = await this.richTextUserChoiceRetrieve();
+    this.choices.richtext = await this.richTextUserChoiceRetrieve();
     for (const key of Object.keys(popoverItems)) {
       const popoverOpt = key as PopoverOpt;
-      if (popoverOpt === 'richText' && Catch.browser().name !== 'firefox') {
-        continue; // richText not deployed to Chrome yet, for now only allow firefox
+      if (popoverOpt === 'richtext' && !this.view.debug && Catch.browser().name !== 'firefox') {
+        continue; // richtext not deployed to Chrome yet, for now only allow firefox
```
Comment (tomholub) [inline extension/chrome/elements/composer/composer-send-btn-popover.ts]: Update comment, mention what debug is for

File: extension/chrome/elements/composer/formatters/encrypted-mail-msg-formatter.ts [hunk 11]
```diff
@@ -35,10 +35,10 @@ export class EncryptedMsgMailFormatter extends BaseMailFormatter implements Mail
     this.armoredPubkeys = armoredPubkeys;
   }
 
-  public sendableMsg = async (newMsg: NewMsgData, signingPrv?: OpenPGP.key.Key): Promise<SendableMsg> => {
+  public sendableMsg = async (newMsg: NewMsgData, signingPrv?: OpenPGP.key.Key, validate: boolean = true): Promise<SendableMsg> => {
```
Comment (tomholub) [inline extension/chrome/elements/composer/formatters/encrypted-mail-msg-formatter.ts]: This value is superfluous, never used

File: extension/js/common/core/pgp-msg.ts [hunk 12]
```diff
@@ -380,6 +380,13 @@ export class PgpMsg {
     return decryptedContent;
   }
 
+  public static createPgpMimeAtts = (content: string) => {
```
Comment (tomholub) [inline extension/js/common/core/pgp-msg.ts]: Should probably be moved to appropriate formatter

File: extension/types/squire.d.ts [hunk 13]
```diff
@@ -30,7 +30,7 @@ export declare class SquireEditor {
   getHTML(): string;
   setHTML(html: string): SquireEditor;
   getSelectedText(): string;
-  insertImage(image: ArrayBuffer, imageAttributes: any): HTMLElement;
+  insertImage(image: ArrayBuffer | string, imageAttributes: any): HTMLElement;
```
Comment (tomholub) [inline extension/types/squire.d.ts]: Investigate, maybe we can stick to `ArrayBuffer`

Comment (tomholub) [reply extension/types/squire.d.ts]: Alternatively send the update upstream

File: test/source/browser/controllable.ts [hunk 14]
```diff
@@ -301,6 +301,25 @@ abstract class ControllableBase {
...
+      const text = await this.read(selector, true);
+      presentForMS += sleepMS;
+      if (!text.includes(content)) {
+        presentForMS = 0;
+      }
+      if (presentForMS === shouldBePresentMS) {
+        return;
+      }
+    }
+    throw new Error('Error!!!');
```
Comment (tomholub) [inline test/source/browser/controllable.ts]: This error is horribly undescriptive

File: test/source/mock/google/google-data.ts [hunk 15]
```diff
@@ -37,16 +97,27 @@ export class GoogleData {
     }
   }
 
-  public storeSentMessage = (parsedMail: ParsedMail): string => {
+  public storeSentMessage = (parsedMail: ParsedMail, base64Msg: string): string => {
+    let attId = '';
+    let signatureAtt: Attachment | undefined;
+    if (parsedMail.attachments?.length) {
```
Comment (tomholub) [inline test/source/mock/google/google-data.ts]: Should probably store all sent attachments

File: test/source/mock/google/google-data.ts [hunk 16]
```diff
@@ -37,16 +97,27 @@ export class GoogleData {
...
+    }
     const barebonesGmailMsg: GmailMsg = { // todo - could be improved - very barebones
       id: `msg_id_${Util.lousyRandom()}`,
       threadId: null, // tslint:disable-line:no-null-keyword
       historyId: '',
       labelIds: ['SENT' as GmailMsg$labelId],
       payload: {
         headers: [{ name: 'Subject', value: parsedMail.subject }],
-        body: { data: parsedMail.text, attachmentId: '', size: parsedMail.text.length }
+        body: { data: parsedMail.text, attachmentId: attId, size: parsedMail.text?.length }
```
Comment (tomholub) [inline test/source/mock/google/google-data.ts]: This looks suspicious - if `parsedMail.text` can be undefined, that means `data` and `size` can remain undefined which seems weird. Will investigate

File: test/source/mock/google/google-endpoints.ts [hunk 17]
```diff
@@ -70,7 +70,8 @@ export const mockGoogleEndpoints: HandlersDefinition = {
         verificationStatus: 'accepted'
       }];
       if (acct === 'flowcrypt.compatibility@gmail.com') {
-        sendAs[0].signature = 'The best footer ever!';
+        // eslint-disable-next-line max-len
+        sendAs[0].signature = '<div dir="ltr">The best footer ever!<br><img src="https://www.smallbizgenius.net/wp-content/uploads/2019/11/smallbizgenius_favicon.png" alt="Image result for small image"><br></div>';
```
Comment (tomholub) [inline test/source/mock/google/google-endpoints.ts]: Should not use a real 3rd party image

File: test/source/mock/google/google-endpoints.ts [hunk 18]
```diff
@@ -252,8 +263,8 @@ const validateMimeMsg = async (acct: string, mimeMsg: ParsedMail, threadId?: str
       throw new HttpClientErr(`Error: Incorrect Email Alias. Should be 'flowcryptcompatibility@gmail.com'. Current '${mimeMsg.from.value[0].address}'`);
     }
   }
-  if (!mimeMsg.text) {
-    throw new HttpClientErr('Error: Message body is required', 400);
+  if (!mimeMsg.text && !mimeMsg.attachments?.length) {
```
Comment (tomholub) [inline test/source/mock/google/google-endpoints.ts]: To review further

File: test/source/mock/google/strategies/send-message-strategy.ts [hunk 19]
```diff
@@ -7,15 +7,23 @@ import { GoogleData } from '../google-data';
 import { HttpClientErr } from '../../lib/api';
 import { PgpMsg } from '../../../core/pgp-msg';
 
+// TODO: Make a better structure of ITestMsgStrategy. Because this class doesn't test anything, it only saves message in the Mock
+class SaveMessageInStorageStrategy implements ITestMsgStrategy {
```
Comment (tomholub) [inline test/source/mock/google/strategies/send-message-strategy.ts]: I mildly remember a similar method before, will have a look

Comment (tomholub) [reply test/source/mock/google/strategies/send-message-strategy.ts]: I think I refactored it away that time

File: test/source/mock/google/strategies/send-message-strategy.ts [hunk 20]
```diff
@@ -126,12 +134,15 @@ export class TestBySubjectStrategyContext {
       this.strategy = new MessageWithFooterTestStrategy();
     } else if (subject.includes('PWD encrypted message')) {
       this.strategy = new PwdEncryptedMessageTestStrategy();
+    } else if (subject.includes('Test Sending Encrypted Message With Image') ||
+      subject.includes('Test Sending Signed Message With Image')) {
+      this.strategy = new SaveMessageInStorageStrategy();
```
Comment (tomholub) [inline test/source/mock/google/strategies/send-message-strategy.ts]: Todo - give this a second look

File: test/source/tests/page_recipe/compose-page-recipe.ts [hunk 21]
```diff
@@ -46,15 +46,15 @@ export class ComposePageRecipe extends PageRecipe {
...
+    sendingOpt: { encrypt?: boolean, sign?: boolean, richtext?: boolean } = {}, // undefined means leave default
     windowType: 'new' | 'reply' = 'new'
   ) {
     await Util.sleep(0.5);
     await ComposePageRecipe.fillRecipients(composePageOrFrame, recipients, windowType);
     if (subject) {
       await composePageOrFrame.click('@input-subject');
       await Util.sleep(1);
-      await composePageOrFrame.type('@input-subject', `Automated puppeteer test: ${subject}`);
+      await composePageOrFrame.type('@input-subject', subject);
```
Comment (tomholub) [inline test/source/tests/page_recipe/compose-page-recipe.ts]: Why was this changed?

File: test/source/tests/tests/compose.ts [hunk 22]
```diff
@@ -533,6 +560,29 @@ export const defineComposeTests = (testVariant: TestVariant, testWithNewBrowser:
 
 };
 
+const testImagesWhenSendingMessages = async (t: AvaContext, browser: BrowserHandle, sendingType: 'encrypt' | 'sign') => {
```
Comment (tomholub) [inline test/source/tests/tests/compose.ts]: This method should be renamed, but not sure what to yet

File: test/source/util/parse.ts [hunk 23]
```diff
@@ -21,7 +20,8 @@ const strictParse = async (source: string): Promise<ParseMsgResult> => {
     throw new Error('ThreadId property doesn\'t exist');
   }
   if (lines[6] === 'Content-Type: message/rfc822' &&
-        lines[7] === 'Content-Transfer-Encoding: base64' && lines[9]) {
+    lines[7] === 'Content-Transfer-Encoding: base64' && lines[9]) {
```
Comment (tomholub) [inline test/source/util/parse.ts]: The indent here looks wrong

File: extension/chrome/elements/composer/composer-footer.ts (hunk 9 above)
Comment (tomholub) [reply extension/chrome/elements/composer/composer-footer.ts]: More importantly, investigate resulting formatting

**Comment by tomholub** [comment]
+5 hours

**Comment by tomholub** [comment]
Yesterday +5 hours

File: test/source/mock/google/strategies/send-message-strategy.ts (hunk 19 above)
Comment (tomholub) [reply test/source/mock/google/strategies/send-message-strategy.ts]: We should not define this as `SaveMessageInStorageStrategy`. Instead, all sent messages should be saved in storage.

Also, each test should have its own storage so that messages sent from one test don't interfere with messages in other tests. Right now all tests share storage. Also, each storage should be wiped between test retries.

**Comment by tomholub** [comment]
Today +1 hour
